op-service/txmgr/txmgr.go

```diff
@@ -350,57 +334,103 @@ func (m *SimpleTxManager) sendTx(ctx context.Context, tx *types.Transaction) (*t
+		return m.l.New("hash", tx.Hash(), "nonce", tx.Nonce(), "gasTipCap", tx.GasTipCap(), "gasFeeCap", tx.GasFeeCap())
+	}
+	l := updateLogFields(tx)
+
+	l.Info("Publishing transaction")
 	t := time.Now()
-	err := m.backend.SendTransaction(cCtx, tx)
-	sendState.ProcessSendError(err)
 
-	// Properly log & exit if there is an error
-	if err != nil {
+	for {
+		if bumpFees {
+			newTx, err := m.increaseGasPrice(ctx, tx)
+			if err != nil {
+				l.Error("unable to increase gas", "err", err)
+				m.metr.TxPublished("bump_failed")
+				return tx
+			}
+			tx = newTx
+			sendState.bumpCount++
+			l = updateLogFields(tx)
+		}
+		bumpFees = true // bump fees next loop
+
+		if sendState.IsWaitingForConfirmation() {
+			// there is a chance the previous tx goes into "waiting for confirmation" state
+			// during the increaseGasPrice call. In some (but not all) cases increaseGasPrice
+			// will error out during gas estimation. In either case we should continue waiting
+			// rather than resubmit the tx.
+			return tx
+		}
+
+		cCtx, cancel := context.WithTimeout(ctx, m.cfg.NetworkTimeout)
+		err := m.backend.SendTransaction(cCtx, tx)
+		cancel()
+		sendState.ProcessSendError(err)
+
+		if err == nil {
+			break
+		}
+
 		switch {
+		case errStringMatch(err, txpool.ErrReplaceUnderpriced):
+			l.Warn("transaction replacement is underpriced", "err", err)
+			m.metr.TxPublished("tx_replacement_underpriced")
+			continue // retry with fee bump
+		case errStringMatch(err, txpool.ErrUnderpriced):
+			l.Warn("transaction is underpriced", "err", err)
+			m.metr.TxPublished("tx_underpriced")
+			continue // retry with fee bump
 		case errStringMatch(err, core.ErrNonceTooLow):
-			log.Warn("nonce too low", "err", err)
+			l.Warn("nonce too low", "err", err)
 			m.metr.TxPublished("nonce_to_low")
 		case errStringMatch(err, context.Canceled):
 			m.metr.RPCError()
-			log.Warn("transaction send cancelled", "err", err)
+			l.Warn("transaction send cancelled", "err", err)
 			m.metr.TxPublished("context_cancelled")
 		case errStringMatch(err, txpool.ErrAlreadyKnown):
-			log.Warn("resubmitted already known transaction", "err", err)
+			l.Warn("resubmitted already known transaction", "err", err)
 			m.metr.TxPublished("tx_already_known")
-		case errStringMatch(err, txpool.ErrReplaceUnderpriced):
-			log.Warn("transaction replacement is underpriced", "err", err)
-			m.metr.TxPublished("tx_replacement_underpriced")
-		case errStringMatch(err, txpool.ErrUnderpriced):
-			log.Warn("transaction is underpriced", "err", err)
-			m.metr.TxPublished("tx_underpriced")
 		default:
 			m.metr.RPCError()
-			log.Error("unable to publish transaction", "err", err)
+			l.Error("unable to publish transaction", "err", err)
 			m.metr.TxPublished("unknown_error")
 		}
-		return
+
+		// on non-underpriced error return immediately; will retry on next resubmission timeout
+		return tx
 	}
-	m.metr.TxPublished("")
 
+	m.metr.TxPublished("")
```

Comment:
The contract of this function is pretty messy; I wonder if it's worth breaking it into two, one which just tries to publish the transaction to the mempool (returning the tx if so) or returns error, and then move the functionality for waiting for it to be mined elsewhere.

If one of those non-expected (non-fee bumping) errors occurred it seems the correct behavior may be to abort entirely rather than wait and try to resubmit exactly as we might do waiting for a successfully published transaction to confirm?

Reply:
Agree, refactored to split publish from wait.

I want to avoid bringing in error handling modifications outside of the underpriced errors.